Building fails on FreeBSD

**Describe the bug**
I am trying to build quickDocs on FreeBSD (12-CURRENT), the build fails with type errors.
```
[3/11] Compiling Vala source ../src/Applica.../PackageList.vala ../src/Widgets/View.vala.
FAILED: src/src@@com.github.mdh34.quickdocs@exe/Application.c src/src@@com.github.mdh34.quickdocs@exe/Constants.c src/src@@com.github.mdh34.quickdocs@exe/MainWindow.c src/src@@com.github.mdh34.quickdocs@exe/Services/Downloader.c src/src@@com.github.mdh34.quickdocs@exe/Widgets/Package.c src/src@@com.github.mdh34.quickdocs@exe/Widgets/PackageList.c src/src@@com.github.mdh34.quickdocs@exe/Widgets/View.c 
valac -C --debug --vapidir /usr/home/dhananjay/code/public/quickDocs/vapi --pkg webkit2gtk-4.0 --pkg libdevhelp-3.0 --pkg libarchive --pkg granite --pkg gtk+-3.0 --pkg gee-0.8 --color=always --directory src/src@@com.github.mdh34.quickdocs@exe --basedir ../src ../src/Application.vala ../src/Constants.vala ../src/MainWindow.vala ../src/Services/Downloader.vala ../src/Widgets/Package.vala ../src/Widgets/PackageList.vala ../src/Widgets/View.vala
../src/Widgets/View.vala:28.17-28.28: warning: `null' incompatible with return type `Gtk.Widget'
                return null;
                ^^^^^^^^^^^^
../src/MainWindow.vala:69.27-69.47: warning: Dh.BookManager.new has been deprecated since 3.26
../src/MainWindow.vala:70.13-70.28: warning: Dh.BookManager.populate has been deprecated since 3.26
../src/Services/Downloader.vala:49.24-49.89: error: 1 extra arguments for `Archive.Result Archive.Read.read_data_block (out weak uint8[], out Archive.int64_t)'
                while (reader.read_data_block (out buffer, out buffer_length, out offset) != Archive.Result.EOF) {
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
../src/Services/Downloader.vala:50.44-50.49: error: Argument 1: Cannot convert from `void*' to `unowned uint8[]'
                    disk.write_data_block (buffer, buffer_length, offset);
                                           ^^^^^^
Compilation failed: 2 error(s), 3 warning(s)
ninja: build stopped: subcommand failed.
```

**To Reproduce**
1. Clone master and try building.

**Expected behavior**
Should build

**System Information**
 - Distribution: FreeBSD 12-CURRENT
 - Vala: Vala 0.42.1
 - Desktop Environment: [Xmonad]
 - Installation Method: [From Source.]

--- a/src/Services/Downloader.vala
+++ b/src/Services/Downloader.vala
@@ -44,11 +44,20 @@ namespace Downloader {
             void* buffer = null;
             Posix.off_t offset;
 
+#if ARCHIVE_332
             if (entry.size () > 0) {
                 while (reader.read_data_block (out buffer, out offset) != Archive.Result.EOF) {
                     disk.write_data_block (buffer, offset);
                 }
             }
+#else
+            size_t buffer_length;
+            if (entry.size () > 0) {
+                while (reader.read_data_block (out buffer, out buffer_length, out offset) != Archive.Result.EOF) {
+                    disk.write_data_block (buffer, buffer_length, offset);
+                }
+            }
+#endif
         }
     }
 

--- a/src/Services/Downloader.vala
+++ b/src/Services/Downloader.vala
@@ -41,16 +41,17 @@ namespace Downloader {
                 continue;
             };
 
-            void* buffer = null;
             Posix.off_t offset;
 
 #if ARCHIVE_332
+            uint8[] buffer;
             if (entry.size () > 0) {
                 while (reader.read_data_block (out buffer, out offset) != Archive.Result.EOF) {
                     disk.write_data_block (buffer, offset);
                 }
             }
 #else
+            void* buffer = null;
             size_t buffer_length;
             if (entry.size () > 0) {
                 while (reader.read_data_block (out buffer, out buffer_length, out offset) != Archive.Result.EOF) {

--- a/src/Services/Downloader.vala
+++ b/src/Services/Downloader.vala
@@ -42,12 +42,11 @@ namespace Downloader {
             };
 
             void* buffer = null;
-            size_t buffer_length;
             Posix.off_t offset;
 
             if (entry.size () > 0) {
-                while (reader.read_data_block (out buffer, out buffer_length, out offset) != Archive.Result.EOF) {
-                    disk.write_data_block (buffer, buffer_length, offset);
+                while (reader.read_data_block (out buffer, out offset) != Archive.Result.EOF) {
+                    disk.write_data_block (buffer, offset);
                 }
             }
         }

--- a/src/meson.build
+++ b/src/meson.build
@@ -1,3 +1,6 @@
+if dependency('libarchive').version().version_compare('>=3.3.2')
+    add_project_arguments('-ARCHIVE_332')
+endif
 executable(
     meson.project_name(),
     'Application.vala',

--- a/src/meson.build
+++ b/src/meson.build
@@ -1,5 +1,5 @@
 if dependency('libarchive').version().version_compare('>=3.3.2')
-    add_project_arguments('-ARCHIVE_332')
+    add_project_arguments('--define=ARCHIVE_332', language: 'vala')
 endif
 executable(
     meson.project_name(),

